### DISCUSSION TITLE
Latency Encoding

### DISCUSSION THREAD

**Discussion by Dola47** [jeshraghian/snntorch#36] [Q&A] [answered]
Hi @jeshraghian,

Thanks a lot for your response.

Here it is once again because I have not precisely understood how to use the Latency encoding.

I present here a snippet that shows how I use the Latency encoding on my dataset, however, I am highly suspicious that it is incorrect, as the latency encoding is getting way better accuracy than the rate encoding. So, please have a look and If I doing the latency encoding incorrectly, then I will be glad if you can provide me with some code that shows how to use it correctly.

```
loss_hist = []
val_loss_hist = []
counter = 0

# Outer training loop
for epoch in range(100):
    minibatch_counter = 0
    train_batch = iter(training_generator)

    # Minibatch training loop
    for data_it, targets_it in train_batch:
        data_it = data_it.to(device)
        targets_it = targets_it.to(device)
        batch_size = data_it.shape[0]
        # Spike generator
        spike_data, spike_targets = spikegen.latency(data_it, targets_it, num_outputs=num_outputs, num_steps=num_steps, tau=5, threshold=0.01,
                                                     linear=True, normalize=True, clip=True)
        
        output, mem_rec = net(spike_data.view(num_steps, batch_size, 95, 32, 32))
        log_p_y = log_softmax_fn(mem_rec)
        loss_val = torch.zeros((1), dtype=dtype, device=device)

        # Sum loss over time steps to perform BPTT
        for step in range(num_steps):
            loss_val += loss_fn(log_p_y[step], targets_it)

        # Gradient calculation
        optimizer.zero_grad()
        loss_val.backward(retain_graph=True)
        nn.utils.clip_grad_norm_(net.parameters(), 1)
        
        # Weight Update
        optimizer.step()

        # Store loss history for future plotting
        loss_hist.append(loss_val.item())

        # Val set
        val_data = itertools.cycle(val_generator)
        valdata_it, valtargets_it = next(val_data)
        valdata_it = valdata_it.to(device)
        valtargets_it = valtargets_it.to(device)
        batch_size = valdata_it.shape[0]
        # Val set spike conversion
        val_spike_data, val_spike_targets = spikegen.latency(valdata_it, valtargets_it, num_outputs=num_outputs, num_steps=num_steps, tau=5, threshold=0.01,
                                                             linear=True, normalize=True, clip=True)
        
        # Val set forward pass
        val_output, val_mem_rec = net(val_spike_data.view(num_steps, batch_size, 95, 32, 32))

        # Val set loss
        log_p_yval = log_softmax_fn(val_mem_rec)
        log_p_yval = log_p_yval.sum(dim=0)
        loss_val_val = loss_fn(log_p_yval, valtargets_it)
        val_loss_hist.append(loss_val_val.item())

        # Print val/train loss/accuracy
        if counter % 50 == 0:
              train_printer()
        minibatch_counter += 1
        counter += 1

loss_hist_true_grad = loss_hist
val_loss_hist_true_grad = val_loss_hist
```
Also just in case if you wonder how my labels look like. I have 8 classes, and I know which parts of my dataset belong to which class. Accordingly, my labels are defined as the following:
```
y_labels=np.zeros((1438))
y_labels[0:180] = 0
y_labels[180:360] = 1
y_labels[360:540] = 2
y_labels[540:720] = 3 
y_labels[720:900] = 4
y_labels[900:1079] = 5
y_labels[1079:1258] = 6
y_labels[1258:1438] = 7
```
In consequence, the `spike_data`  is just tensors consisting of ones and zeros, and `spike_targets`  is their corresponding labels.

Looking forward to your response as soon as possible.

Thanks.

**Comment by jeshraghian** [comment]
Hey @Dola47,

- I see you're applying latency to the input, though you are training your network to create a rate-coded output. There's nothing wrong with this approach, but I just want to make sure you are aware. 

- Are you using the values in `y_labels` as the targets in `targets_it`?

- What does `95` represent in the third dimension?

**Reply by Dola47** [reply]
Hi @jeshraghian ,

> I see your applying latency to the input, though you are training your network to create a rate-coded output. There's nothing wrong with this approach, but I just want to make sure you are aware.

That is not actually my aim, I need to have everything Latency encoded. I see then that I do not know how to insert or use my labels when I latency encodes my data. It would be helpful then if you can show me how to do that. Can you provide some modification to the above code to have the data, and labels latency encoded and trained upon that?

Moreover, can you provide some clarification on how could it be an acceptable approach to latency encode the input data, and keep the network learning rate encoded output? It does not seem logical to me. Also, I got very good results out of it and the networks have been able to learn all of the training data, but I can not see how the network is really working and how it learned anything :).

> Are you using the values in y_labels as the targets in targets_it?

That's correct.

> What does 95 represent in the third dimension?

I am considering it as my number of channels, my input dataset is of the shape of 95, 32, 32.

Looking forward to your response.

Thanks.

**Comment by jeshraghian** [comment, marked as answer]
> That is not actually my aim, I need to have everything Latency encoded, then I do not know how to insert or use my labels when I latency encodes my data. It would be helpful then if you can show me how to do that. Can you provide some modification to the above code to have the data, and labels latency encoded and trained upon that?

Unfortunately I don't have the time to write your training loop, though I may include it in future tutorials. 

Until then, here are some issues with your code that suggest you're training a rate coded output, and not latency: 

`loss_fn(log_p_y[step], targets_it)` - your target is encouraging your network to fire at every single time step. This is effectively rate coding.
 
If you want latency coding at your output, your loss must only promote the correct class to fire first. I am assuming `loss_fn` is implementing cross entropy or negative log likelihood. You should see that your current target is trying make your correct class fire at every single time step. There is no official best practice to this yet. I would recommend using mean square error on the membrane. Then your target needs to be a tensor that has the correct class firing first, and incorrect classes firing later (or not firing at all). These can be achieved natively in numpy and PyTorch. 

Your next problem will be `train_printer()`. If you haven't modified it from the code in the tutorial, you are measuring accuracy by counting the total number of output spikes. This is incorrect - latency coding should only count the class that fires first.

**Reply by Dola47** [reply]
You are correct with all of your points.

I can say that I understand the problem right now, however, I do not know where all of the modifications should happen. I would be thankful if you can provide modification to my wrong steps as much as you can. Moreover, can you sum up in some points what should be modified precisely ( for example the targets should be defined in another way, another loss function, another counting of correct classification, etc ....)?

Thanks.

**Reply by Dola47** [reply]
What if I kept the train_printer() and the loss function, and only used latency encoding on my inputs (as shown in my very first comment). I see then that the latency encoding is nothing but just a step to convert my input into tensors of zeroes and ones, and as you mentioned earlier the network is still working in a rate encoded way. Does that mean then that the accuracy I am getting is incorrect? And that the network is not really working correctly?

**Reply by jeshraghian** [reply]
> Does that mean then that the accuracy I am getting is incorrect? And that the network is not really working correctly?

If your aim is to create a rate-coded output, then your approach seems generally correct.
However, a possible error in your code you should look into further. 

Check the lines in `print_batch_accuracy`:

```
train_acc = print_batch_accuracy(spike_data, spike_targets, train=True)
val_acc = print_batch_accuracy(val_spike_data, val_spike_targets, train=False)
```

You are applying `spike_targets` and `val_spike_targets`. 
But in your training loop, you are using non-spiking targets. That might be the source of your error.
Your code may need to look something like this:

```
train_acc = print_batch_accuracy(spike_data, targets_it, train=True)
val_acc = print_batch_accuracy(val_spike_data, valtargets_it, train=False)
```

**Reply by Dola47** [reply]
Hi @jeshraghian,

Well first of all I am sorry that this discussion went long, it seems that you are quite busy and I hope I am not causing any problems :).

I will sum up my comment into 3 parts.

Part 1:

- I see that there is no difference between Latency encoding and Rate encoding for the network working prespective except that in rate encoding we are encouraging the network to fire each time step `loss_fn(log_p_y[step], targets_it)`, and that we count the number of spikes for accuracy. If I am correct then using only the Latency encoder to encode our data, then keeping the network working in a rate encoding manner as I am doing right now should be an acceptable approach. Because both `spikegen.rate` and `spikegen.latency` do nothing except converting our input data into tensors full of zeroes and ones and they will generate tensor of the same shape, only in the latency encoding case the converted data will have fewer spikes than the rate encoded data. As the latency encoding will let every pixel spike once, however, the rate encoder will generate spike trains for each pixel. Accordingly, I see that the network can also use the latency encoded input data to learn rate encoded output.
- Please confirm or correct my understanding here.

Part2:

> Check the lines in print_batch_accuracy:
          train_acc = print_batch_accuracy(spike_data, spike_targets, train=True)
          val_acc = print_batch_accuracy(val_spike_data, val_spike_targets, train=False)
You are applying spike_targets and val_spike_targets.
But in your training loop, you are using non-spiking targets. That might be the source of your error.

- Just to note, I am following your unreleased tutorial about Convnets, so this might be an error that you have not corrected yet. But, from my point of view, I see that there is no difference between for example `spike_targets` which is the output of `spikegen.latency`, and `targets_it` which is submitted to the targets argument of the `spikegen.latency`. Because according to the current code the target is not converted to spikes!! It is just the same tensor. So, using `targets_it` and `valtargets_it` instead of `spike_targets `, and `val_spike_targets` should make no difference. And even in the unreleased code for Convnets you used `targets_it`  for the training loss and used  `test_spike_targets`  for the test loss, and for the accuracies, you used normally the `spike_targets` and `test_spike_targets` .

Part3:

- I think it will be very very helpful to have as soon as possible one example of how to use the Latency encoding and delta encoding end-to-end in our network. and to have the Convnet tutorial officially released. 

Part4:

- Really thank you once again for replying and answering my questions.

**Reply by jeshraghian** [reply]
> Part 1. Please confirm or correct my understanding here.

[Tutorial 1](https://snntorch.readthedocs.io/en/latest/tutorials/tutorial_1.html) has a deep explanation of Rate coding and latency coding input data. 

As for a network's output layer, rate encoding encourages the correct class to have the highest firing frequency. Latency encoding encourages the correct class to fire first. 

> Just to note, I am following your unreleased tutorial about Convnets

Yep, as it is unreleased it'll obviously have a lot of bugs in it. So be wary of learning from it. 

If `spike_targets` and `targets_it` are the same, then your output is not being latency encoded correctly. I can't tell if it's an snnTorch or a coding problem, but if it's an snnTorch error then 

For latency coding of your targets, please refer to tutorial 1 which will teach you how to convert tensors into latency encoded data. 
However, you may discover that PyTorch's cross-entropy loss is not compatible with latency encoding, i.e., it cannot set all labels to 0 for a time-step. 
I expect Mean Square Error Loss to be more suitable here. 
If this is unclear to you, then I recommend referring to the PyTorch documentation on loss functions.

> I think it will be very very helpful to have as soon as possible one example of how to use the Latency encoding and delta encoding end-to-end in our network. and to have the Convnet tutorial officially released.

I do this on a voluntary basis & it's not my main job. 
So it'll obviously take a while as I'm currently the only maintainer of the tutorials. 
Thanks for your patience!

**Reply by Dola47** [reply]
Hi @jeshraghian,

Let's sum it up.

I am doing latency encoding as in Tutorial 1. Exactly as 
```
spike_data, spike_targets = spikegen.latency(data_it, targets_it, num_outputs=num_outputs, num_steps=num_steps, tau=5, threshold=0.01, linear=True, normalize=True, clip=True)
```

Here is how my `spike_data `looks, at least for me it is correct, as it is generating only one spike out of each pixel.
![image](https://user-images.githubusercontent.com/45386621/116682152-2a0f9500-a9ae-11eb-95e3-7e5d50d5f900.png)

As already mentioned `spike_targets ` is the same as the input `targets_it`. Either because of an error in the `spikgen.latency` or it is working correctly but because I am not doing time-varying targets or one-hot encoding. 
Anyway, I think for my current implementation it is not important to have the correct targets from the `spikgen.latency`, because I am gonna do rate decoding and learning anyhow. 

So, right now I have one spike out of each pixel, I am gonna motivate the correct class to spike at each time step through `loss_fn(log_p_y[step], targets_it)` and `NLLLoss` and `SoftMax`. When it comes to calculating the accuracy I will count the no. of spikes I have for each class from the last spiking neuron. Here is a slide showing the accuracy from one of the batches I have

![image](https://user-images.githubusercontent.com/45386621/116684233-026dfc00-a9b1-11eb-8c77-0d6cb90b43cb.png)

Actually, this approach of applying latency encoding to the inputs, and keep the network doing rate encode learning and spike accuracy counting is getting good results.

So, the question is, is something wrong with this approach, or it is from the engineering perspective acceptable. Even though we agree that there is no time decoding happening?

Looking forward to your response.

Thanks.

**Reply by jeshraghian** [reply]
> As already mentioned spike_targets is the same as the input targets_it. Either because of an error in the spikgen.latency or it is working correctly

Thanks for bringing this to my attention. I'll test to see what bugs I can catch in `spikegen.latency`.
Feel free to open an issue copying in the code you use so I can reproduce the bug (including how you determine `spike_targets` and `targets_it` are the same). That would be super helpful to me while I try and clean that code up!

> So, the question is, is something wrong with this approach, or it is from the engineering perspective acceptable. Even though we agree that there is no time decoding happening?

Your approach looks good to me.
You have latency encoded the input (such that the most intense values spike first). 
Then you are passing the same target into NLL + softmax for every time step, and so your network is using rate-coding to make a prediction. In terms of deep learning, it's totally acceptable!

Deep learning doesn't constrain how you mix + match latency/rate/delta coded _inputs_ to rate/latency trained _outputs_. 
I hope to do a demo of latency coded outputs soon, as it's not directly compatible with NLLLoss / CrossEntropyLoss. Aim to combine that with the conv tutorial.

**Reply by Dola47** [reply]
Hi @jeshraghian,

Thanks a lot for your response.

To check the error in `spikegen.latency` I just run Tutorial 1. Until step 2.1 where you start to call the `spikegen.rate`.

Instead, I called `spikegen.latency` . And here is the output 

![image](https://user-images.githubusercontent.com/45386621/116767494-b6629c00-aa30-11eb-8f25-d04e78b819a4.png).

You can see that the spiked targets are exactly the same as the targets_it, just it is the case for `spikegen.rate` .

> I hope to do a demo of latency coded outputs soon, as it's not directly compatible with NLLLoss / CrossEntropyLoss. Aim to combine that with the conv tutorial.

That's gonna be perfect, hope they will be released soon as April is already over :) :).

**Reply by jeshraghian** [reply]
> You can see that the spiked targets are exactly the same as the targets_it, just it is the case for spikegen.rate .

Gotcha! Yeah as you spotted yourself, `targets__it == spike_targets_latency` because the `one_hot_encoding` and `time_varying_targets` haven't been set to `True`. However, this doesn't seem to be very intuitive. 
I'll think of a way to improve this!

**Reply by Dola47** [reply]
Yes! But we do not usually need our targets to be one-hot encoded, and if we just used `time_varying_targets` we will get the same target only repeated along the time domain, then we only need to change `loss_fn(log_p_y[step], targets_it)` to `loss_fn(log_p_y[step], targets_it[step])` which actually do nothing! 

And all of that I think does not have any relation to that currently the targets are not temporal encoded in the way that it motivates the correct class to fire first and the incorrect class to fire late. So, as far as I understand you are right that the latency encoding of the targets should be modified.

**Reply by jeshraghian** [reply]
> But we do not usually need our targets to be one-hot encoded

Yep - in PyTorch you're exactly right! For NLLLoss and CrossEntropyLoss, we use indexes rather than one-hot-encodings.
So the initial purpose of one-hot-encodings was to apply them to MSELoss. I.e., each neuron generates its own MSELoss, and the one-hot-encodings could be used to encourage the membrane potential to reach '1' for the correct class, and '0' for the incorrect class. 

In some cases, you might want the membrane to be '1.05' or some tiny amount above the threshold to guarantee a spike occurs. And you might want the incorrect classes to be '0.25' or some other non-zero value, so that the output neurons are somewhat ready to fire if the input changes (instead of encouraging the weights to be pushed close to zero). 

On the other hand, in my tests with MNIST, I've noticed that the membrane of the correct class is often explodes when using NLLLoss + softmax. Even though the accuracy results are acceptable, such large membrane potentials are probably not the most efficient way. 

Appreciate your feedback - I will think on a way to efficiently incorporate all of this into the next release :)

**Reply by Dola47** [reply]
Hi @jeshraghian,

One more notice.  I am trying to save the output spikes of the network for each batch. However, I am usually getting the error of  
```
Warning: the spikes outside of the range of num_steps have been clipped.
 Setting ``normalize = True`` or increasing ``num_steps`` can fix this.
```

However, I am setting normalize to True, and clip to True as the following:
```
for epoch in trange(50):
    minibatch_counter = 0

    # Minibatch training loop
    for (data_it, targets_it) in tqdm(training_generator):
        data_it = data_it.to(device)
        targets_it = targets_it.to(device)
        batch_size = data_it.shape[0]
        # Spike generator
        spike_data, spike_targets = spikegen.latency(data_it, targets_it, num_outputs=num_outputs, num_steps=num_steps, tau=5, threshold=0.01,
                                                     linear=True, normalize=True, clip=True)
        
        output, mem_rec = net(spike_data.view(num_steps, batch_size, 95, 32, 32))

        # Retun the spikes and the voltages for each batch
        spikes_hist.append(output)
        voltages_hist.append(mem_rec)
```
Any ideas on how to save the output spikes, because I need to test if the network is working correctly or not?

Moreover, for the membrane of the correct class that explodes when you use NLLLoss + SoftMax. I think you are talking out the `mem_rec` of the network. Can you specify what value you mean by saying that it is exploding, is it over 1? Because normally for me it is under 1. And it is not usually the correct class that comes with the maximum voltage btw in my case.

Thanks a lot Jason.

**Reply by jeshraghian** [reply]
> Any ideas on how to save the output spikes, because I need to test if the network is working correctly or not?

The warning just indicates that when you apply latency coding, some of your inputs spike outside of the range of ``num_steps``. 
E.g., an input of `1` will fire at the first time step. But an input of `0` will take much longer. 

However, if your normalize is set to True, then all of those final spikes should be squashed into the range of ``num_steps``, and therefore, suppress the warning message.
This looks like an error on snnTorch's side - not yours. Are you able to open an issue up so I can replicate your error?

>  I think you are talking out the mem_rec of the network

That's correct; although this is very data-dependent. 
When I trained on MNIST, I found the membrane to go much larger than 1 using NLLLoss + SoftMax.

> it is not usually the correct class that comes with the maximum voltage btw in my case.

This is a tougher question to answer. You can specify the correct class by using max voltage, although I use the spike count.
The issues might also be arising from imbalanced datasets, poor initialization parameters, poor network parameters, or simply that BPTT is not the best training method for your data, or any other number of possibilities. 
The SNN research community is still trying to solve this problem for natively trained SNNs.

**Reply by Dola47** [reply]
Sorry for the late reply.

> This looks like an error on snnTorch's side - not yours. Are you able to open an issue up so I can replicate your error?

I tried to replicate the error on MNIST, but the warning does not show up, but It threw a memory error. So, I solved the memory error as well on my dataset and then the warning is gone. I do not know why the warning raised, but anyways the problem is resolved now.

**Reply by jeshraghian** [reply]
> but It threw a memory error

That most likely happened for large datasets/large time-steps and would require either smaller batches, smaller `num_steps`, or multiple GPUs.

>  I do not know why the warning raised, but anyways the problem is resolved now.

Glad to hear it. I have almost finished an updated form of target encoding functions.

**Reply by Dola47** [reply]
Hi @jeshraghian,

Any updates regarding the time to first spike target encoding and MSE Loss and decoding?

Thanks.

**Reply by jeshraghian** [reply]
The latest changes include time-to-first-spike encoding for inputs using `spikegen.rate(input, num_steps=num_steps, first_spike_time=fst)` and for targets using `spikegen.targets_rate(targets, num_classes=num_classes, num_steps=num_steps, first_spike_time=fst)`. 

Take a look at the documentation - there is a lot of info there on how to use it. 

As a sample of how to apply MSE Loss:

```
loss_fn = nn.MSELoss()

targets_mem = spikegen.targets_rate(targets, num_classes=10, on_target=1.05, off_target=0.25)  # returns one-hot encoding where incorrect classes are 0.25, correct classes are 1.05
# You could of course leave on_target and off_target to default values for a simple one-hot-encoding, but I like to encourage the membrane to overshoot the threshold by a little bit to guarantee firing, and I like to ensure incorrect classes are semi-ready to fire.

# feedforward pass
spk_rec, mem_rec = net(data_it.view(batch_size, -1))

# initialize loss to 0
loss_val = torch.zeros((1), dtype=dtype, device=device)

# accumulate MSE loss for all time steps
for step in range(num_steps):
    loss_val += loss_fn(mem_rec[step], targets_mem)

loss_val.backward()
```

Note that `targets_mem` is not time-varying, so it is a 2D tensor. 
If you were to modify certain arguments, such as `first_spike_time`, `correct_rate`, `incorrect_rate`, then it would be time-varying. You would then iterate through `targets_mem` by modifying the loss calculation to:

```
for step in range(num_steps):
    loss_val += loss_fn(mem_rec[step], targets_mem[step])
```

**Reply by Dola47** [reply]
@jeshraghian 

Thanks a lot, it is becoming clearer now.

What about how to calculate the accuracy, for the rate encoding we were actually counting the no of spikes. If we now latency encode the targets and the data, how you gonna consider the class that spiked first and classify it as the correct class?

Looking forward to your response.

Thanks.

**Reply by jeshraghian** [reply]
I don't have a code snippet for that yet, so you'll need to try implementing this yourself.

But conceptually, if your output `spk_rec` is of size `num_steps x batch_size x num_classes` then you need to find the index for where the first (if any) '1' occurs in the `num_steps` dimension. 

The predicted class would then be the minimum index across the `num_classes` dimension.

**Comment by Wang-jun-yu** [comment]
Do you have any example code that uses Latency Encoding to recognize the MNIST dataset? If so, could you please provide a link to it?

**Reply by Dola47** [reply]
Hi @Wang-jun-yu,

Sorry for the late reply.

I remember that I faced a problem in calculating the loss, and the network was not trained appropriately. Accordingly, I have not looked further into it. 

Unfortunately, I do not have an implementation on MNIST. It is just a quick trial on a private dataset. 

Thanks.

**Reply by Spice-monkey** [reply]
hey, I am also searching for the examples for Latency Encoding. If you find the relevant code, please contact with me.Thanks.

**Reply by SevillaFe** [reply]
Hello @jeshraghian , good afternoon,
I would like to know if there are any examples of using latency encoding applied to regression tasks or if you know of any material that could help.
Thanks to all
